### PR TITLE
!str introduce StreamletLogic trait

### PR DESCRIPTION
**Pull request by hepin1989**
I think this would be better.

### CODE REVIEW

File: core/cloudflow-streamlets/src/main/scala/cloudflow/streamlets/StreamletLogic.scala [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+   * Returns the [[StreamletContext]] in which this [[StreamletLogic]] is run. It can only be accessed when the streamlet is run.
+   */
+  def context: Context
+
+  /**
+   * Java API
+   *
+   * Returns the [[StreamletContext]] in which this [[StreamletLogic]] is run. It can only be accessed when the streamlet is run.
+   */
+  def getContext(): Context = context
```
Comment (hepin1989) [inline core/cloudflow-streamlets/src/main/scala/cloudflow/streamlets/StreamletLogic.scala]: for java api

File: core/cloudflow-akka-tests/src/test/java/cloudflow/akkastream/javadsl/AkkaStreamletTest.java [hunk 2]
```diff
@@ -105,8 +105,9 @@ public StreamletShape shape() {
       return StreamletShape.createWithInlets(inlet).withOutlets(outlet);
     }
 
-    public StreamletLogic createLogic() {
-      return new StreamletLogic(getStreamletContext()) {
+    public AkkaStreamletLogic createLogic() {
+      final AkkaStreamletContext context = getStreamletContext();
```
Comment (RayRoestenburg) [inline core/cloudflow-akka-tests/src/test/java/cloudflow/akkastream/javadsl/AkkaStreamletTest.java]: This does not seem necessary?

File: core/cloudflow-flink-tests/src/test/java/cloudflow/flink/javadsl/FlinkProcessor.java [hunk 3]
```diff
@@ -45,15 +45,16 @@
   // Step 3: Provide custom implementation of `FlinkStreamletLogic` that defines
   //         the behavior of the streamlet
   @Override public FlinkStreamletLogic createLogic() {
-    return new FlinkStreamletLogic(getStreamletContext()) {
+    final FlinkStreamletContext context =  getStreamletContext();
```
Comment (RayRoestenburg) [inline core/cloudflow-flink-tests/src/test/java/cloudflow/flink/javadsl/FlinkProcessor.java]: This also does not seem necessary, if I revert this change, everything compiles?

Comment (hepin1989) [reply core/cloudflow-flink-tests/src/test/java/cloudflow/flink/javadsl/FlinkProcessor.java]: done

File: core/cloudflow-akka-tests/src/test/java/cloudflow/akkastream/javadsl/AkkaStreamletTest.java (hunk 2 above)
Comment (hepin1989) [reply core/cloudflow-akka-tests/src/test/java/cloudflow/akkastream/javadsl/AkkaStreamletTest.java]: true

File: core/cloudflow-flink/src/main/scala/cloudflow/flink/FlinkStreamlet.scala [hunk 4]
```diff
@@ -270,7 +269,9 @@ abstract class FlinkStreamlet extends Streamlet with Serializable {
  *  }
  * }}}
  */
-abstract class FlinkStreamletLogic(implicit context: FlinkStreamletContext) extends Serializable {
+abstract class FlinkStreamletLogic(implicit val context: FlinkStreamletContext) extends StreamletLogic[FlinkStreamletContext] {
+  override def getContext(): FlinkStreamletContext = super.getContext()
```
Comment (RayRoestenburg) [inline core/cloudflow-flink/src/main/scala/cloudflow/flink/FlinkStreamlet.scala]: Is this one necessary?

File: examples/sensor-data-java/project/cloudflow-plugins.sbt [hunk 5]
```diff
@@ -2,4 +2,4 @@
 //
 resolvers += "Akka Snapshots" at "https://repo.akka.io/snapshots/"
 
-addSbtPlugin("com.lightbend.cloudflow" % "sbt-cloudflow" % "1.3.0-M1")
+addSbtPlugin("com.lightbend.cloudflow" % "sbt-cloudflow" % "1.3.0-M4")
```
Comment (RayRoestenburg) [inline examples/sensor-data-java/project/cloudflow-plugins.sbt]: 👍

File: core/cloudflow-flink/src/main/scala/cloudflow/flink/FlinkStreamlet.scala (hunk 4 above)
Comment (hepin1989) [reply core/cloudflow-flink/src/main/scala/cloudflow/flink/FlinkStreamlet.scala]: yes.

Comment (RayRoestenburg) [reply core/cloudflow-flink/src/main/scala/cloudflow/flink/FlinkStreamlet.scala]: Ah, I see, it is for java compatibility. Ok.

**Comment by RayRoestenburg** [comment]
Thanks @hepin1989 !
